widget/text.go

```diff
@@ -341,11 +341,57 @@ func splitLines(text []rune) [][2]int {
 	return append(lines, [2]int{low, length})
 }
 
+// newBinarySearch accepts a function that checks if the text width between the start and end index less the maximum width
+// newBinarySearch returns a function that accepts the start and end rune index and returns the index of rune
+// located as close to the maximum line width as possible
+func newBinarySearch(lessMaxWidth func(int, int) bool) func(low int, maxHigh int) int {
+	return func(low int, maxHigh int) int {
+		if low >= maxHigh {
+			return low
+		}
+		if lessMaxWidth(low, maxHigh) {
+			return maxHigh
+		}
+		high := low
+		delta := maxHigh - low
+		for delta > 0 {
+			delta /= 2
+			if lessMaxWidth(low, high+delta) {
+				high += delta
+			}
+		}
+		for (high < maxHigh) && lessMaxWidth(low, high+1) {
+			high++
+		}
+		return high
+	}
+}
+
+// findSpaceIndex accepts a slice of runes and the last rune index
+// findSpaceIndex returns the index of the last space in the text, or fallback if there are no spaces
+func findSpaceIndex(text []rune, fallback int) int {
+	curIndex := fallback
+	for ; curIndex >= 0; curIndex-- {
+		if unicode.IsSpace(text[curIndex]) {
+			break
+		}
+	}
+	if curIndex < 0 {
+		return fallback
+	}
+	return curIndex
+}
+
 // lineBounds accepts a slice of runes, a wrapping mode, a maximum line width and a function to measure line width.
 // lineBounds returns a slice containing the start and end indicies of each line with the given wrapping applied.
 func lineBounds(text []rune, wrap fyne.TextWrap, maxWidth int, measurer func([]rune) int) [][2]int {
+
+	binarySearch := newBinarySearch(func(low int, high int) bool {
```

Comment:
BenchmarkText_lineBounds_WrapOff is faster if you move this initialization after line 396